_source/_posts/2021-12-16-cloud-native-java-microservices-with-istio.md

```diff
@@ -0,0 +1,580 @@
+```
+
+This will create the applications with all its entities and specified configurations. You should be able to see the gateway application in action by running the below command on the **store** folder. You will be redirected to Keycloak for login.
+
+```bash
+$ docker-compose -f src/main/docker/keycloak.yml up -d # starts keycloak in daemon mode
+$ ./gradlew # starts the spring boot application
+```
+
+You can find a [sample application on GitHub](https://github.com/oktadev/okta-java-spring-k8s-istio-microservices-example).
+
+## Configure Okta as Identity Provider
+
+The applications are configured by default to use Keycloak as the identity provider since we choose OAuth2 as authentication mechanism. However, you can switch to any other OAuth2 provider by editing the **application.yml** file for each of the applications.
+
+Let's see how to configure Okta as the identity provider. You may wonder? Why Okta? Keycloak is great for development and testing but its not robust for production as you would need to manage it and keep it updated and so on. Okta on the other hand lets you worry about your business needs by keeping your identity requirements always on and secure.
+
+### Create an Okta application
+
+{% include setup/cli.md type="jhipster" %}
+
+### Configure OpenID Connect settings to use Okta
+
+Open `src/main/resources/config/application.yml` file for each application and update `issuer-uri`, `client-id` and `client-secret` under the `spring.security.oauth2.client` section as shown below with values you obtained during previous step.
+
+```yml
+spring:
+  ...
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://{yourOktaDomain}/oauth2/default
+        registration:
+          oidc:
+            client-id: {yourClientId}
+            client-secret:
+            scope: openid,profile,email
+```
+
+> **Note**: For security reasons we will not set the client-secret on this file and instead use environment variable, `SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET` or Kubernetes secrets. You can also skip updating the application.yml files altogether and use environment variable or Kubernetes secrets for all the OIDC configurations.
+
+You can run `./gradlew` on the store application to test the Okta login. Make sure to set the client secret environment variable first.
+
+## Create a GKE Cluster and Install Istio
+
+In order to deploy the stack to Google Kubernetes Engine, we need to create a cluster and install Istio. So lets begin by creating a cluster using Google CLoud SDK.
+
+### Create a cluster
+
+Make sure you are logged into the gcloud CLI and run the below command to create a GKE cluster.
+
+```bash
+# set region and zone
+$ gcloud config set compute/region europe-west1
+$ gcloud config set compute/zone europe-west1-b
+# Create a project and enable container APIs
+$ gcloud projects create jhipster-demo # You need to also enable billing via GUI
```

Comment:
You might want to mention that you'll need another name here since "jhipster-demo" is already in use.